In [2]:
#Importation des librairies et chemins
import pandas as pd
import numpy as np
from pathlib import Path


#Affichage des colonnes 
pd.set_option("display.max_columns", 100)
pd.set_option("display.max_rows", 100)

DATA_PATH = Path("../data/consolidation-etalab-schema-irve-statique-v-2.3.1-20260625.csv")

In [3]:
#Chargement des données
df_bronze = pd.read_csv(
    DATA_PATH,
    low_memory=False
)

print("Lignes :", df_bronze.shape[0])
print("Colonnes :", df_bronze.shape[1])

Lignes : 227232
Colonnes : 52


df_bronze c'est la donnée brute qu'on ne modifie jamais

In [4]:
#Copie de dz_bronze
df = df_bronze.copy()
df.head()

,nom_amenageur,siren_amenageur,contact_amenageur,nom_operateur,contact_operateur,telephone_operateur,nom_enseigne,id_station_itinerance,id_station_local,nom_station,implantation_station,adresse_station,code_insee_commune,coordonneesXY,nbre_pdc,id_pdc_itinerance,id_pdc_local,puissance_nominale,prise_type_ef,prise_type_2,prise_type_combo_ccs,prise_type_chademo,prise_type_autre,gratuit,paiement_acte,paiement_cb,paiement_autre,tarification,condition_acces,reservation,horaires,accessibilite_pmr,restriction_gabarit,station_deux_roues,raccordement,num_pdl,date_mise_en_service,observations,date_maj,cable_t2_attache,last_modified,datagouv_dataset_id,datagouv_resource_id,datagouv_organization_or_owner,created_at,consolidated_longitude,consolidated_latitude,consolidated_code_postal,consolidated_commune,consolidated_is_lon_lat_correct,consolidated_is_code_insee_verified,consolidated_is_code_insee_modified
0,ChargePoint,NaN,info@chargepoint.com,ChargePoint,info@chargepoint.com,NaN,ACU_Poste_De_Garde_Haguenau,ATHTBE1004017,ATHTBE1004017,ACU_Poste_De_Garde_Haguenau,Parking privé à usage public,"93 route de Bitche, 67506 Haguenau Cedex",NaN,"[7.76269400,48.82561300]",4,ATHTBE1004017,ATHTBE1004017,22.0,false,true,false,false,false,false,true,true,NaN,NaN,Accès réservé,false,24/7,Accessibilité inconnue,inconnu,false,NaN,NaN,2020-07-25,EF connector is available at the site separately,2026-06-24,false,2026-06-24T21:00:21.100000+00:00,64060c2ac773dcf3fabbe5d2,b11113db-875d-41c7-8673-0cf8ad43e917,eco-movement,2023-06-28T11:46:08.539000+00:00,7.762694,48.825613,NaN,NaN,False,False,False
1,ChargePoint,NaN,info@chargepoint.com,ChargePoint,info@chargepoint.com,NaN,ACU_Poste_De_Garde_Haguenau,ATHTBE1004018,ATHTBE1004018,ACU_Poste_De_Garde_Haguenau,Parking privé à usage public,"93 route de Bitche, 67506 Haguenau Cedex",NaN,"[7.76269400,48.82561300]",4,ATHTBE1004018,ATHTBE1004018,22.0,false,true,false,false,false,false,true,true,NaN,NaN,Accès réservé,false,24/7,Accessibilité inconnue,inconnu,false,NaN,NaN,2020-07-25,EF connector is available at the site separately,2026-06-24,false,2026-06-24T21:00:21.100000+00:00,64060c2ac773dcf3fabbe5d2,b11113db-875d-41c7-8673-0cf8ad43e917,eco-movement,2023-06-28T11:46:08.539000+00:00,7.762694,48.825613,NaN,NaN,False,False,False
2,ChargePoint,NaN,info@chargepoint.com,ChargePoint,info@chargepoint.com,NaN,ACU_Poste_De_Garde_Haguenau,ATHTBE1004019,ATHTBE1004019,ACU_Poste_De_Garde_Haguenau,Parking privé à usage public,"93 route de Bitche, 67506 Haguenau Cedex",NaN,"[7.76269400,48.82561300]",4,ATHTBE1004019,ATHTBE1004019,11.0,false,true,false,false,false,false,true,true,NaN,NaN,Accès réservé,false,24/7,Accessibilité inconnue,inconnu,false,NaN,NaN,2020-07-25,EF connector is available at the site separately,2026-06-24,false,2026-06-24T21:00:21.100000+00:00,64060c2ac773dcf3fabbe5d2,b11113db-875d-41c7-8673-0cf8ad43e917,eco-movement,2023-06-28T11:46:08.539000+00:00,7.762694,48.825613,NaN,NaN,False,False,False
3,ChargePoint,NaN,info@chargepoint.com,ChargePoint,info@chargepoint.com,NaN,ACU_Poste_De_Garde_Haguenau,ATHTBE1004020,ATHTBE1004020,ACU_Poste_De_Garde_Haguenau,Parking privé à usage public,"93 route de Bitche, 67506 Haguenau Cedex",NaN,"[7.76269400,48.82561300]",4,ATHTBE1004020,ATHTBE1004020,11.0,false,true,false,false,false,false,true,true,NaN,NaN,Accès réservé,false,24/7,Accessibilité inconnue,inconnu,false,NaN,NaN,2020-07-25,EF connector is available at the site separately,2026-06-24,false,2026-06-24T21:00:21.100000+00:00,64060c2ac773dcf3fabbe5d2,b11113db-875d-41c7-8673-0cf8ad43e917,eco-movement,2023-06-28T11:46:08.539000+00:00,7.762694,48.825613,NaN,NaN,False,False,False
4,ChargePoint,NaN,info@chargepoint.com,ChargePoint,info@chargepoint.com,NaN,HAG_P22_Slave_3,ATHTBE1004957,ATHTBE1004957,HAG_P22_Slave_3,Parking privé à usage public,"Route de Bitche 93, 67500 Haguenau",NaN,"[7.76184100,48.82704000]",8,ATHTBE1004957,ATHTBE1004957,22.0,false,true,false,false,false,false,true,true,NaN,NaN,Accès réservé,true,24/7,Accessibilit

1 - Profiling et audit qualité

In [5]:
#Vérification des colonnes
colonnes = pd.DataFrame({
    "colonne": df.columns,
    "type": df.dtypes.astype(str),
    "valeurs_non_nulles": df.notna().sum().values,
    "valeurs_nulles": df.isna().sum().values,
    "taux_nulles_%": (df.isna().mean().values * 100).round(2)
})

colonnes.sort_values("taux_nulles_%", ascending=False)

,colonne,type,valeurs_non_nulles,valeurs_nulles,taux_nulles_%
observations,observations,str,46520,180712,79.53
tarification,tarification,str,56345,170887,75.20
num_pdl,num_pdl,str,115972,111260,48.96
cable_t2_attache,cable_t2_attache,str,120491,106741,46.97
consolidated_code_postal,consolidated_code_postal,float64,134362,92870,40.87
date_mise_en_service,date_mise_en_service,str,134961,92271,40.61
siren_amenageur,siren_amenageur,float64,136247,90985,40.04
raccordement,raccordement,str,147925,79307,34.90
id_pdc_local,id_pdc_local,str,148554,78678,34.62
consolidated_commune,consolidated_commune,str,150839,76393,33.62


In [6]:
#colonnes clés
colonnes_cles = [
    "id_pdc_itinerance",
    "id_station_itinerance",
    "date_maj",
    "last_modified",
    "date_mise_en_service",
    "puissance_nominale",
    "condition_acces",
    "coordonneesXY",
    "consolidated_longitude",
    "consolidated_latitude",
    "consolidated_code_postal",
    "consolidated_commune",
    "consolidated_is_lon_lat_correct"
]

df[colonnes_cles].head()

,id_pdc_itinerance,id_station_itinerance,date_maj,last_modified,date_mise_en_service,puissance_nominale,condition_acces,coordonneesXY,consolidated_longitude,consolidated_latitude,consolidated_code_postal,consolidated_commune,consolidated_is_lon_lat_correct
0,ATHTBE1004017,ATHTBE1004017,2026-06-24,2026-06-24T21:00:21.100000+00:00,2020-07-25,22.0,Accès réservé,"[7.76269400,48.82561300]",7.762694,48.825613,NaN,NaN,False
1,ATHTBE1004018,ATHTBE1004018,2026-06-24,2026-06-24T21:00:21.100000+00:00,2020-07-25,22.0,Accès réservé,"[7.76269400,48.82561300]",7.762694,48.825613,NaN,NaN,False
2,ATHTBE1004019,ATHTBE1004019,2026-06-24,2026-06-24T21:00:21.100000+00:00,2020-07-25,11.0,Accès réservé,"[7.76269400,48.82561300]",7.762694,48.825613,NaN,NaN,False
3,ATHTBE1004020,ATHTBE1004020,2026-06-24,2026-06-24T21:00:21.100000+00:00,2020-07-25,11.0,Accès réservé,"[7.76269400,48.82561300]",7.762694,48.825613,NaN,NaN,False
4,ATHTBE1004957,ATHTBE1004957,2026-06-24,2026-06-24T21:00:21.100000+00:00,2024-04-30,22.0,Accès réservé,"[7.76184100,48.82704000]",7.761841,48.827040,NaN,NaN,False


In [7]:
#Contrôle des dimensions globales
audit_global = pd.DataFrame({
    "indicateur": [
        "Nombre de lignes",
        "Nombre de colonnes",
        "Nombre de cellules",
        "Nombre total de valeurs manquantes",
        "Taux global de valeurs manquantes (%)",
        "Nombre de lignes entièrement dupliquées"
    ],
    "valeur": [
        df.shape[0],
        df.shape[1],
        df.shape[0] * df.shape[1],
        df.isna().sum().sum(),
        round(df.isna().sum().sum() / (df.shape[0] * df.shape[1]) * 100, 2),
        df.duplicated().sum()
    ]
})

audit_global

,indicateur,valeur
0,Nombre de lignes,227232.00
1,Nombre de colonnes,52.00
2,Nombre de cellules,11816064.00
3,Nombre total de valeurs manquantes,1449476.00
4,Taux global de valeurs manquantes (%),12.27
5,Nombre de lignes entièrement dupliquées,0.00


In [8]:
# Analyse des identifiants répétés

# Nombre total de lignes du dataset
nb_lignes = len(df)

# Nombre d'identifiants de points de charge distincts
# dropna=True : les valeurs manquantes ne sont pas comptées
nb_id_pdc_uniques = df["id_pdc_itinerance"].nunique(dropna=True)

# Nombre d'identifiants manquants
nb_id_pdc_manquants = df["id_pdc_itinerance"].isna().sum()

# Nombre de lignes concernées par un identifiant répété
# keep=False marque toutes les occurrences comme doublons potentiels
nb_lignes_id_repetes = (
    df.duplicated(
        subset=["id_pdc_itinerance"],
        keep=False
    )
).sum()

# Nombre de répétitions supplémentaires
# C'est ce calcul qui correspond aux 67 094 mentionnés dans le sujet
nb_doublons = (
    df["id_pdc_itinerance"]
    .duplicated()
    .sum()
)

# Création du tableau récapitulatif
analyse_grain = pd.DataFrame({
    "indicateur": [
        "Nombre total de lignes",
        "Nombre d'id_pdc_itinerance uniques",
        "Nombre d'id_pdc_itinerance manquants",
        "Nombre de lignes avec id_pdc_itinerance répété",
        "Nombre de doublons (répétitions supplémentaires)"
    ],
    "valeur": [
        nb_lignes,
        nb_id_pdc_uniques,
        nb_id_pdc_manquants,
        nb_lignes_id_repetes,
        nb_doublons
    ]
})

analyse_grain

,indicateur,valeur
0,Nombre total de lignes,227232
1,Nombre d'id_pdc_itinerance uniques,160138
2,Nombre d'id_pdc_itinerance manquants,0
3,Nombre de lignes avec id_pdc_itinerance répété,130964
4,Nombre de doublons (répétitions supplémentaires),67094


2 - Analyse du grain et des doublons

## Analyse d'un identifiant répété

In [9]:
# Comptage du nombre d'occurrences par identifiant

repetition_id_pdc = (
    df["id_pdc_itinerance"]
    .value_counts()
    .reset_index()
)

repetition_id_pdc.columns = [
    "id_pdc_itinerance",
    "nombre_occurrences"
]

# Affichage des identifiants les plus fréquents

repetition_id_pdc.head(20)

,id_pdc_itinerance,nombre_occurrences
0,Non concerné,118
1,FRLIBE3126085,4
2,FRA68P680210015,3
3,FRALLEGO0002061,3
4,FRALLEGO0002062,3
5,FRALLEGO0020201,3
6,FRALLEGO0020202,3
7,FRALLEGO0020211,3
8,FRALLEGO0020212,3
9,FRALLEGO0020231,3


## Comparaison détaillée d'un identifiant répété

In [10]:
# Sélection d'un identifiant apparaissant plusieurs fois

id_test = "FRLIBE3126085"

# Affichage des lignes correspondantes

df[
    df["id_pdc_itinerance"] == id_test
][[
    "id_pdc_itinerance",
    "id_station_itinerance",
    "nom_station",
    "adresse_station",
    "puissance_nominale",
    "date_maj",
    "last_modified",
    "created_at"
]].sort_values("date_maj")

,id_pdc_itinerance,id_station_itinerance,nom_station,adresse_station,puissance_nominale,date_maj,last_modified,created_at
105427,FRLIBE3126085,FRLIBE3126085,Residence Carouge,34 Rue de l’Orge 91220 BrŽtigny-sur-Orge,22.0,2023-10-11,2026-01-09T13:01:51.509000+00:00,2026-01-09T13:01:51.318000+00:00
105428,FRLIBE3126085,FRLIBE3126085,Residence Carouge,34 Rue de l‚Äö√Ñ√¥Orge 91220 Brétigny-sur-Orge,22.0,2023-10-11,2025-11-24T21:47:43.507000+00:00,2025-11-24T21:47:43.223000+00:00
105425,FRLIBE3126085,FRLIBE3126085,Résidence Carouge,34 rue de l'Orge,7.4,2023-10-21,2024-12-06T06:47:29.008000+00:00,2023-10-21T22:11:40.997000+00:00
105426,FRLIBE3126085,FRLIBE3126085,Residence Carouge,34 Rue de l’Orge 91220 Brétigny-sur-Orge,22.0,2023-11-10,2024-12-06T06:47:51.017000+00:00,2023-11-13T09:31:03.098000+00:00


1- La puissance nominale est différente pour chaque id donc ce n'est pas des doublons(Cohérence)
2- Problème d'encodage détecté(Validité)
3- Les dates sont différentes, colonnes (created_at, last_modified) ce qui correspondent sûrement 
à plusieurs livraisons successives

## Analyses des répétitions exactes

In [11]:
# Nombre de lignes totalement identiques
nb_lignes_identiques = df.duplicated().sum()

print(
    "Nombre de lignes strictement identiques :",
    nb_lignes_identiques
)

Nombre de lignes strictement identiques : 0


### Conclusion analyse doublon

L'analyse montre qu'aucune ligne du dataset n'est strictement identique à une autre.

Les 67 094 répétitions observées concernent uniquement l'identifiant `id_pdc_itinerance`.

L'étude d'un exemple réel a montré que les lignes associées possèdent des différences de contenu (dates, puissance, libellés), ce qui suggère qu'il ne s'agit pas de doublons techniques mais potentiellement de plusieurs livraisons ou versions d'un même point de charge.

Aucune suppression ne sera effectuée à ce stade sans analyse métier complémentaire.

In [12]:
## Analyse des colonnes temporelles
df[[
    "date_maj",
    "last_modified",
    "created_at"
]].head(10)

,date_maj,last_modified,created_at
0,2026-06-24,2026-06-24T21:00:21.100000+00:00,2023-06-28T11:46:08.539000+00:00
1,2026-06-24,2026-06-24T21:00:21.100000+00:00,2023-06-28T11:46:08.539000+00:00
2,2026-06-24,2026-06-24T21:00:21.100000+00:00,2023-06-28T11:46:08.539000+00:00
3,2026-06-24,2026-06-24T21:00:21.100000+00:00,2023-06-28T11:46:08.539000+00:00
4,2026-06-24,2026-06-24T21:00:21.100000+00:00,2023-06-28T11:46:08.539000+00:00
5,2026-06-24,2026-06-24T21:00:21.100000+00:00,2023-06-28T11:46:08.539000+00:00
6,2026-06-24,2026-06-24T21:00:21.100000+00:00,2023-06-28T11:46:08.539000+00:00
7,2026-06-24,2026-06-24T21:00:21.100000+00:00,2023-06-28T11:46:08.539000+00:00
8,2026-06-24,2026-06-24T21:00:21.100000+00:00,2023-06-28T11:46:08.539000+00:00
9,2026-06-24,2026-06-24T21:00:21.100000+00:00,2023-06-28T11:46:08.539000+00:00


In [13]:
# Vérification du nombre de valeurs différentes
# dans les colonnes temporelles

df[[
    "date_maj",
    "last_modified",
    "created_at"
]].nunique()

date_maj         1220
last_modified     806
created_at        806
dtype: int64

In [14]:
##Analyse temporelle des identifiants répétés
# Exemple sur l'identifiant déjà étudié

df[
    df["id_pdc_itinerance"] == "FRLIBE3126085"
][[
    "id_pdc_itinerance",
    "date_maj",
    "last_modified",
    "created_at"
]].sort_values("date_maj")


,id_pdc_itinerance,date_maj,last_modified,created_at
105427,FRLIBE3126085,2023-10-11,2026-01-09T13:01:51.509000+00:00,2026-01-09T13:01:51.318000+00:00
105428,FRLIBE3126085,2023-10-11,2025-11-24T21:47:43.507000+00:00,2025-11-24T21:47:43.223000+00:00
105425,FRLIBE3126085,2023-10-21,2024-12-06T06:47:29.008000+00:00,2023-10-21T22:11:40.997000+00:00
105426,FRLIBE3126085,2023-11-10,2024-12-06T06:47:51.017000+00:00,2023-11-13T09:31:03.098000+00:00


In [15]:
# Conversion en datetime

df["date_maj_dt"] = pd.to_datetime(
    df["date_maj"],
    errors="coerce"
)

df["last_modified_dt"] = pd.to_datetime(
    df["last_modified"],
    errors="coerce"
)

df["created_at_dt"] = pd.to_datetime(
    df["created_at"],
    errors="coerce"
)

# Résumé statistique

resume_dates = pd.DataFrame({
    "min": [
        df["date_maj_dt"].min(),
        df["last_modified_dt"].min(),
        df["created_at_dt"].min()
    ],
    "max": [
        df["date_maj_dt"].max(),
        df["last_modified_dt"].max(),
        df["created_at_dt"].max()
    ],
    "nb_valeurs": [
        df["date_maj_dt"].count(),
        df["last_modified_dt"].count(),
        df["created_at_dt"].count()
    ]
},
index=[
    "date_maj_dt",
    "last_modified_dt",
    "created_at_dt"
])

resume_dates

,min,max,nb_valeurs
date_maj_dt,2012-09-10 00:00:00,2026-12-30 00:00:00,226940
last_modified_dt,2024-02-09 06:22:40.878000+00:00,2026-06-24 21:00:21.100000+00:00,193355
created_at_dt,2021-05-10 21:01:29.234000+00:00,2026-06-24 15:14:32.641000+00:00,227232


## Impact métier des répétitions d'identifiants

In [16]:
# Nombre total de lignes
nb_lignes = len(df)

# Nombre de bornes distinctes
nb_bornes = df["id_pdc_itinerance"].nunique()

# Nombre de répétitions supplémentaires
nb_repetitions = nb_lignes - nb_bornes

impact_grain = pd.DataFrame({
    "Indicateur": [
        "Nombre total de lignes",
        "Nombre de bornes distinctes",
        "Répétitions supplémentaires"
    ],
    "Valeur": [
        nb_lignes,
        nb_bornes,
        nb_repetitions
    ]
})

impact_grain

,Indicateur,Valeur
0,Nombre total de lignes,227232
1,Nombre de bornes distinctes,160138
2,Répétitions supplémentaires,67094


## Journal de décisions - Grille d'audit qualité

### Problème

67 094 répétitions de l'identifiant `id_pdc_itinerance`.

### Options envisagées

- Supprimer toutes les répétitions avec `drop_duplicates()`
- Conserver uniquement la version la plus récente
- Analyser le grain avant toute suppression

### Choix retenu

Analyser le grain avant toute suppression.

### Justification métier

L'étude d'un identifiant répété a montré que les lignes ne sont pas identiques et peuvent correspondre à plusieurs versions ou livraisons d'un même point de charge.

Le sujet indique explicitement que le grain du fichier est potentiellement « point de charge × livraison ». Une suppression automatique risquerait donc de supprimer des données valides.

3. Nettoyage et enrichissement

In [17]:
# Création d'une copie de travail pour le nettoyage
# Les colonnes temporaires créées pendant l'audit
# (date_maj_dt, last_modified_dt, created_at_dt)
# ne sont pas conservées dans la couche Silver.

df_silver = df.copy()

df_silver = df_silver.drop(
    columns=[
        "date_maj_dt",
        "last_modified_dt",
        "created_at_dt"
    ],
    errors="ignore"
)

print("Dataset Silver créé")
print("Lignes :", df_silver.shape[0])
print("Colonnes :", df_silver.shape[1])

Dataset Silver créé
Lignes : 227232
Colonnes : 52


## Analyse des anomalies d'encodage

In [18]:
# Recherche de caractères suspects pouvant révéler
# un problème d'encodage (UTF-8 / Latin-1 / CP1252)

colonnes_textes = df_silver.select_dtypes(include="object").columns

caracteres_suspects = ["Ã", "Â", "�", "Ä", "Ö", "Û"]

resultats_encodage = []

for col in colonnes_textes:

    masque = df_silver[col].astype(str).str.contains(
        "|".join(caracteres_suspects),
        regex=True,
        na=False
    )

    nb = masque.sum()

    if nb > 0:
        resultats_encodage.append(
            [col, nb]
        )

encodage_df = pd.DataFrame(
    resultats_encodage,
    columns=[
        "colonne",
        "nb_valeurs_suspectes"
    ]
).sort_values(
    "nb_valeurs_suspectes",
    ascending=False
)

encodage_df

C:\Users\Dell\AppData\Local\Temp\ipykernel_2656\4182633727.py:4: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  colonnes_textes = df_silver.select_dtypes(include="object").columns


,colonne,nb_valeurs_suspectes
3,nom_station,535
4,adresse_station,453
7,observations,415
1,nom_enseigne,105
0,nom_amenageur,53
6,restriction_gabarit,12
2,id_station_local,3
5,tarification,1


### Exemples de valeurs présentant des anomalies d'encodage

In [19]:
# Affichage de quelques exemples dans les colonnes les plus touchées

for col in ["nom_station", "adresse_station"]:

    print(f"\n===== {col} =====")

    exemples = df_silver[
        df_silver[col].astype(str).str.contains(
            "Ã|Â|�|Ä|Ö|Û",
            regex=True,
            na=False
        )
    ][col].drop_duplicates().head(10)

    for valeur in exemples:
        print(valeur)


===== nom_station =====
CLUB MED - GRAND MASSIF SAMÖENS
HÔTEL DES CHÂTEAUX - AZAY
IBIS BUDGET CHÂTELLERAULT
B&B CHÂTELLERAULT
CHÂTEAU-GAILLARD - Rue Jean Mermoz
CHÂLONS - GHV
CAP KOAD - PLÉCHÂTEL
CHÂTILLON-EN-DIOIS-Parking Etrée Ouest
LAVOÛTE SUR LOIRE - Place du Pont Neuf
CHÂTEL -Parking télécabine du Linga

===== adresse_station =====
 Boulevard Victor Hugo 3, 86100 CHÂTELLERAULT
 Rue Jean Mermoz, 1500 CHÂTEAU-GAILLARD
 Rue des Viviers, 51000 CHÂLONS-EN-CHAMPAGNE
rue des CentaureÃÅes
6 D103 43800 LAVOÛTE-SUR-LOIRE
357 Route de Pré-la-joux 74390 CHÂTEL
4172 Route de Pré la Joux 74390 CHÂTEL
19 Route Napoléon 38119 PIERRE-CHÂTEL
295 Chemin des Ruisseaux 74390 CHÂTEL
16 Avenue de Valence 26300 CHÂTEAUNEUF-SUR-ISÈRE


In [20]:
# Recherche uniquement des motifs typiques
# d'erreurs UTF-8 / Latin-1

motifs_encodage = ["Ã", "Â", "�"]

for col in ["nom_station", "adresse_station"]:

    print(f"\n===== {col} =====")

    exemples = df_silver[
        df_silver[col].astype(str).str.contains(
            "|".join(motifs_encodage),
            regex=True,
            na=False
        )
    ][col].drop_duplicates().head(20)

    for valeur in exemples:
        print(valeur)


===== nom_station =====
HÔTEL DES CHÂTEAUX - AZAY
IBIS BUDGET CHÂTELLERAULT
B&B CHÂTELLERAULT
CHÂTEAU-GAILLARD - Rue Jean Mermoz
CHÂLONS - GHV
CAP KOAD - PLÉCHÂTEL
CHÂTILLON-EN-DIOIS-Parking Etrée Ouest
CHÂTEL -Parking télécabine du Linga
CHÂTEL - Pre la joux
PIERRE CHÂTEL - Place des Anciens Combattants AFN
CHÂTEL - Route de Meurba
CHÂTEL -  Route de Meurba
CHÂTEAUNEUF SUR ISERE - Quai de l'Isere
CHÂTEAU BERNARD - Parking de l'Eglise
ARÂCHES LA FRASSE - Flaine 2
ARÂCHES-LA-FRASSE -Parking de la Piscine
SAINT PAUL TROIS CHÂTEAUX - Espace Gare
CHÂTEAUDOUBLE - Place de l'eglise
CHÂTILLON SUR CLUSES - Route de Taninges
SAINT PAUL TROIS CHÂTEAUX - Parking Chaussy

===== adresse_station =====
 Boulevard Victor Hugo 3, 86100 CHÂTELLERAULT
 Rue Jean Mermoz, 1500 CHÂTEAU-GAILLARD
 Rue des Viviers, 51000 CHÂLONS-EN-CHAMPAGNE
rue des CentaureÃÅes
357 Route de Pré-la-joux 74390 CHÂTEL
4172 Route de Pré la Joux 74390 CHÂTEL
19 Route Napoléon 38119 PIERRE-CHÂTEL
295 Chemin des Ruisseaux 74390 CHÂT

In [21]:
# Recherche des différentes valeurs de condition d'accès

df_silver["condition_acces"].value_counts(dropna=False).head(30)

condition_acces
Accès libre      194476
Accès réservé     30708
Accčs libre        1807
Accs libre         224
Acc¸s libre          16
Accès libre          1
Name: count, dtype: int64

### Normalisation des valeurs de condition_acces

In [22]:
# Normalisation des variantes d'encodage

mapping_condition_acces = {
    "Accčs libre": "Accès libre",
    "Acc�s libre": "Accès libre",
    "Acc¸s libre": "Accès libre",
    "Accès libre": "Accès libre"
}

df_silver["condition_acces"] = (
    df_silver["condition_acces"]
    .replace(mapping_condition_acces)
)

In [23]:
df_silver["condition_acces"].value_counts(dropna=False)

condition_acces
Accès libre      196299
Accès réservé     30708
Accs libre         224
Accès libre          1
Name: count, dtype: int64

toute valeur de condition_acces qui contient “libre” devient Accès libre, et toute valeur qui contient “réservé” reste Accès réservé.

In [24]:
# Normalisation de la colonne condition_acces
# Objectif : regrouper toutes les variantes corrompues de "Accès libre"

df_silver["condition_acces"] = np.where(
    df_silver["condition_acces"].astype(str).str.contains("libre", case=False, na=False),
    "Accès libre",
    df_silver["condition_acces"]
)

df_silver["condition_acces"] = np.where(
    df_silver["condition_acces"].astype(str).str.contains("réservé|reservé|reserve|réserve", case=False, regex=True, na=False),
    "Accès réservé",
    df_silver["condition_acces"]
)

# Vérification après correction
df_silver["condition_acces"].value_counts(dropna=False)

condition_acces
Accès libre      196524
Accès réservé     30708
Name: count, dtype: int64

### Analyse de la casse des communes

In [25]:
# Analyse des formats utilisés dans les communes

df_silver["consolidated_commune"].dropna().sample(20, random_state=42)

133867                 Saint-Jean
91873                       Lille
131644                      Cergy
129391    Sainte-Marie-aux-Chênes
39836                      Vienne
75212                Petite-Forêt
13441                    Grenoble
210754                   Xertigny
128762                     Évreux
226207                      Genas
5654                     Lanester
106381                    Vivonne
187213                   Malakoff
185939                   Morangis
101476                 Chalezeule
4497          Paray-Vieille-Poste
48568           Saint-Apollinaire
152027                   Poitiers
71927                        Auby
134948                  Perpignan
Name: consolidated_commune, dtype: str

## Normalisons les formats de dates

In [26]:
# Vérification des colonnes contenant des dates

colonnes_dates = [
    "date_mise_en_service",
    "date_maj",
    "created_at",
    "last_modified"
]

for col in colonnes_dates:
    print(f"\n===== {col} =====")
    print(df_silver[col].dropna().head(10))


===== date_mise_en_service =====
0    2020-07-25
1    2020-07-25
2    2020-07-25
3    2020-07-25
4    2024-04-30
5    2024-04-30
6    2024-04-30
7    2024-04-30
8    2024-05-05
9    2024-05-05
Name: date_mise_en_service, dtype: str

===== date_maj =====
0    2026-06-24
1    2026-06-24
2    2026-06-24
3    2026-06-24
4    2026-06-24
5    2026-06-24
6    2026-06-24
7    2026-06-24
8    2026-06-24
9    2026-06-24
Name: date_maj, dtype: str

===== created_at =====
0    2023-06-28T11:46:08.539000+00:00
1    2023-06-28T11:46:08.539000+00:00
2    2023-06-28T11:46:08.539000+00:00
3    2023-06-28T11:46:08.539000+00:00
4    2023-06-28T11:46:08.539000+00:00
5    2023-06-28T11:46:08.539000+00:00
6    2023-06-28T11:46:08.539000+00:00
7    2023-06-28T11:46:08.539000+00:00
8    2023-06-28T11:46:08.539000+00:00
9    2023-06-28T11:46:08.539000+00:00
Name: created_at, dtype: str

===== last_modified =====
0    2026-06-24T21:00:21.100000+00:00
1    2026-06-24T21:00:21.100000+00:00
2    2026-06-24T21:00:

### Conversion des colonnes dates au format datetime

In [27]:
# Conversion des dates

df_silver["date_mise_en_service"] = pd.to_datetime(
    df_silver["date_mise_en_service"],
    errors="coerce"
)

df_silver["date_maj"] = pd.to_datetime(
    df_silver["date_maj"],
    errors="coerce"
)

df_silver["created_at"] = pd.to_datetime(
    df_silver["created_at"],
    errors="coerce"
)

df_silver["last_modified"] = pd.to_datetime(
    df_silver["last_modified"],
    errors="coerce"
)

df_silver[
    [
        "date_mise_en_service",
        "date_maj",
        "created_at",
        "last_modified"
    ]
].dtypes

date_mise_en_service         datetime64[us]
date_maj                     datetime64[us]
created_at              datetime64[us, UTC]
last_modified           datetime64[us, UTC]
dtype: object

### Analyse des puissances nominales

In [28]:
# Statistiques descriptives de la puissance

df_silver["puissance_nominale"].describe()

count    227232.000000
mean        102.212347
std         579.594838
min           0.000000
25%          22.000000
50%          22.000000
75%         100.000000
max      160000.000000
Name: puissance_nominale, dtype: float64

In [29]:
# Les plus petites puissances

df_silver["puissance_nominale"]\
    .sort_values()\
    .head(20)

79069     0.0
165090    0.0
165089    0.0
165088    0.0
118506    0.0
118505    0.0
6046      0.0
6043      0.0
6040      0.0
6037      0.0
6034      0.0
6031      0.0
6028      0.0
6025      0.0
165077    0.0
165076    0.0
165075    0.0
165074    0.0
165073    0.0
165072    0.0
Name: puissance_nominale, dtype: float64

In [30]:
# Les plus grandes puissances

df_silver["puissance_nominale"]\
    .sort_values(ascending=False)\
    .head(20)

107024    160000.0
226854     50000.0
226856     50000.0
107023     22000.0
224747     16000.0
224751     16000.0
224746     12800.0
224750     12800.0
227082      7360.0
226529      7360.0
226530      7360.0
226559      7360.0
227072      7360.0
226519      7360.0
226223      7360.0
226227      7360.0
226228      7360.0
226229      7360.0
226230      7360.0
226238      7360.0
Name: puissance_nominale, dtype: float64

In [31]:
# Comptage des puissances nulles

nb_puissance_zero = (
    df_silver["puissance_nominale"] == 0
).sum()

# Comptage des puissances > 1000
# probablement exprimées en watts

nb_puissance_watts = (
    df_silver["puissance_nominale"] > 1000
).sum()

print("Puissances nulles :", nb_puissance_zero)
print("Puissances > 1000 :", nb_puissance_watts)

Puissances nulles : 5282
Puissances > 1000 : 975


In [32]:
# Valeurs les plus fréquentes parmi les puissances > 1000

df_silver.loc[
    df_silver["puissance_nominale"] > 1000,
    "puissance_nominale"
].value_counts().head(20)

puissance_nominale
7360.0      769
3680.0       90
1242.0       77
1500.0       10
1200.0        9
6900.0        4
1840.0        4
12800.0       2
16000.0       2
50000.0       2
1601.6        1
22000.0       1
160000.0      1
5750.0        1
2530.0        1
3220.0        1
Name: count, dtype: int64

In [33]:
# Flag des puissances suspectes

df_silver["flag_puissance_watts"] = (
    df_silver["puissance_nominale"] > 1000
)

df_silver["flag_puissance_nulle"] = (
    df_silver["puissance_nominale"] == 0
)

print("Puissances > 1000 :", df_silver["flag_puissance_watts"].sum())
print("Puissances nulles :", df_silver["flag_puissance_nulle"].sum())

Puissances > 1000 : 975
Puissances nulles : 5282


Colonne code_postal

In [34]:
df_silver["consolidated_code_postal"].isna().sum()

np.int64(92870)

In [35]:
# Imputation locale des codes postaux manquants
# Principe : si plusieurs lignes ont les mêmes coordonnées GPS,
# et qu'une d'elles possède un code postal, on réutilise ce code postal.

cp_avant = df_silver["consolidated_code_postal"].isna().sum()

mapping_gps_cp = (
    df_silver
    .dropna(subset=[
        "consolidated_latitude",
        "consolidated_longitude",
        "consolidated_code_postal"
    ])
    .drop_duplicates(subset=[
        "consolidated_latitude",
        "consolidated_longitude"
    ])
    .set_index([
        "consolidated_latitude",
        "consolidated_longitude"
    ])["consolidated_code_postal"]
)

masque_cp_manquant = df_silver["consolidated_code_postal"].isna()

df_silver.loc[
    masque_cp_manquant,
    "consolidated_code_postal"
] = df_silver.loc[
    masque_cp_manquant
].set_index([
    "consolidated_latitude",
    "consolidated_longitude"
]).index.map(mapping_gps_cp)

cp_apres = df_silver["consolidated_code_postal"].isna().sum()

print("Codes postaux manquants avant :", cp_avant)
print("Codes postaux manquants après :", cp_apres)
print("Codes postaux imputés :", cp_avant - cp_apres)

Codes postaux manquants avant : 92870
Codes postaux manquants après : 79938
Codes postaux imputés : 12932


Imputation à partir des coordonnées GPS existantes. Lorsque plusieurs bornes partagent les mêmes coordonnées GPS, le code postal d'une borne connue peut être réutilisé pour compléter les autres.

In [36]:
df_silver.shape

(227232, 54)

D'après l'analyse :

le grain métier = id_pdc_itinerance
- plusieurs lignes existent pour un même identifiant
- elles correspondent à des versions successives de la même borne
- j'ai observé des date_maj, created_at et last_modified différents

Donc la règle la plus défendable est :

Pour chaque id_pdc_itinerance,
conserver la ligne la plus récente selon date_maj.

Dédoublonnage des points de charge

In [37]:
# Nombre de lignes avant dédoublonnage

nb_avant = len(df_silver)

# Tri par date de mise à jour
df_silver = df_silver.sort_values("date_maj")

# Conservation de la version la plus récente
df_silver = df_silver.drop_duplicates(
    subset=["id_pdc_itinerance"],
    keep="last"
)

nb_apres = len(df_silver)

print("Lignes avant :", nb_avant)
print("Lignes après :", nb_apres)
print("Lignes supprimées :", nb_avant - nb_apres)

Lignes avant : 227232
Lignes après : 160138
Lignes supprimées : 67094


## Dédoublonnage des points de charge

### Problème observé

Plusieurs lignes partageaient le même identifiant `id_pdc_itinerance`.

### Vérification effectuée

L'analyse des dates a montré que ces lignes correspondaient à différentes versions d'un même point de charge et non à des doublons techniques stricts.

### Options envisagées

1. Conserver toutes les lignes
2. Supprimer arbitrairement les répétitions
3. Conserver la version la plus récente

### Choix retenu

Conserver la ligne la plus récente selon `date_maj`.

### Justification métier

Les répétitions correspondent à l'historique des mises à jour d'un même point de charge.

### Résultat

- Avant : 227232 lignes
- Après : 160138 lignes
- Lignes supprimées : 67094

## Score qualité après nettoyage

In [38]:
# Nombre total de cellules

nb_cellules = (
    df_silver.shape[0]
    * df_silver.shape[1]
)

# Valeurs manquantes

nb_null = df_silver.isna().sum().sum()

# Taux global de nullité

tx_null = round(
    (nb_null / nb_cellules) * 100,
    2
)

# Doublons sur le grain métier

nb_doublons = (
    df_silver["id_pdc_itinerance"]
    .duplicated()
    .sum()
)

score_apres = pd.DataFrame({
    "indicateur": [
        "Nombre de lignes",
        "Nombre de colonnes",
        "Valeurs manquantes",
        "Taux global de nullité (%)",
        "Doublons id_pdc_itinerance"
    ],
    "valeur": [
        df_silver.shape[0],
        df_silver.shape[1],
        nb_null,
        tx_null,
        nb_doublons
    ]
})

score_apres

,indicateur,valeur
0,Nombre de lignes,160138.00
1,Nombre de colonnes,54.00
2,Valeurs manquantes,1143276.00
3,Taux global de nullité (%),13.22
4,Doublons id_pdc_itinerance,0.00
